File: python/y2024/day03.py
from aoc.core import solve, AocInput
from dataclasses import dataclass
from functools import reduce
from typing import Match
import re


@dataclass(frozen=True, slots=True)
class State:
    enabled: bool
    total: int


def instructions(m: Match) -> int:
    return int(m.groups()[0]) * int(m.groups()[1])


def conditional_instructions(state: State, m: Match) -> State:
    match m.group(0):
        case "do()":
            return State(True, state.total)
        case "don't()":
            return State(False, state.total)
        case _ if state.enabled:
            return State(state.enabled, state.total + instructions(m))
        case _:
            return state


def part1(input: AocInput) -> int:
    return sum(instructions(m) for m in re.finditer(r"mul\((\d+),(\d+)\)", input.text))


def part2(input: AocInput) -> int:
    state = reduce(
        conditional_instructions,
        [m for m in re.finditer(r"do\(\)|don't\(\)|mul\((\d+),(\d+)\)", input.text)],
        State(True, 0),
    )
    return state.total


def main() -> None:
    YEAR = "y2024"
    DAY = "Day03"
    print(f"Running year {YEAR.replace('y', '')} day {DAY.replace('Day', '')}")
    solve(YEAR, DAY, part1, 166357705, True)
    solve(YEAR, DAY, part2, 88811886, True)

File: dotnet/2024/AdventOfCode.Y2024/Day03.cs
using System.Text.RegularExpressions;
using AdventOfCode.Lib;
using State = (bool Enabled, long Total);

namespace AdventOfCode.Y2024;

[AocPuzzle(2024, 3, "Mull It Over", "C#")]
public sealed partial class Day03 : IAocDay<long>
{
    public static long Part1(AocInput input) =>
        input.Text
        | (text => ComputerMemory().Matches(text).Sum(Instructions));

    public static long Part2(AocInput input) => 
        input.Text
        | (text => ComputerMemoryWithConditionals().Matches(text)
            .Aggregate(
                new State(true, 0),
                (state, m) => m.Value switch
                {
                    "do()" => state with { Enabled = true },
                    "don't()" => state with { Enabled = false },
                    _ when state.Enabled => state with
                    {
                        Total = state.Total + Instructions(m)
                    },
                    _ => state
                }
            )
        )
        | (state => state.Total);

    private static int Instructions(Match m) =>
        int.Parse(m.Groups[1].Value) * int.Parse(m.Groups[2].Value);

    [GeneratedRegex(@"mul\((\d+),(\d+)\)", RegexOptions.Compiled)]
    private static partial Regex ComputerMemory();

    [GeneratedRegex(@"do\(\)|don't\(\)|mul\((\d+),(\d+)\)", RegexOptions.Compiled)]
    private static partial Regex ComputerMemoryWithConditionals();
}